### ISSUE TITLE
Duplicate Notification on Android Devices 

### ISSUE THREAD

**Issue by hbabhishek** [ParsePlatform/Parse-SDK-Android#173]
Hello Parse Team,

We are sending push notification to Android users based on their Installation class Object Id. But some users are receiving duplicate notifications. 

We checked our server log and it shows that we have sent the notification only one time. 

I have attached the duplicate notification screenshot sent by user with this. 

![screenshot_2015-09-09-11-26-08](https://cloud.githubusercontent.com/assets/14979235/10282816/59661a70-6b9a-11e5-8b6e-4b14aafdc911.png)

I am using following PHP code to initiate parse push for android device user -
 
        function androidParseNotification($device_id = '', $message = 'hi', $extra = array()) {
               /*echo $device_id;
               echo $message;exit;*/
              $url = 'https://api.parse.com/1/push';
              $result = '';
              try {
                   if (empty($device_id)) {
                   throw new Exception("Device token not found..!");
                   }
            if(is_array($device_id))
            $device_id = implode(',', $device_id);
            $device_tokenArr = explode(',', $device_id);

            $APPLICATION_ID = $this->CI->config->item('PARSE_APPLICATION_ID');
            $REST_API_KEY = $this->CI->config->item('PARSE_REST_API_KEY');
            
            if (empty($REST_API_KEY) || empty($APPLICATION_ID)) {
                throw new Exception("Push notification Authorization key not found..!");
            }
            $payLoadArray = $extra;
            $payLoadArray['alert1'] = $message;
            $push_payload = json_encode(array(
                                           "where" => array(
                                        "objectId" => array('$in' => $device_tokenArr),
                                ),
                    "data" => $payLoadArray
            ));
            //pr($push_payload);
            if (!function_exists('curl_version')) {
                throw new Exception("CURL is not installed in this server..!");
            }

            // Open connection
            $rest = curl_init();
            if (!$rest) {
                throw new Exception("CURL intialization fails..!");
            }
            // Set the url, number of POST vars, POST data
            curl_setopt($rest,CURLOPT_URL,$url);
            curl_setopt($rest,CURLOPT_PORT,443);
            curl_setopt($rest,CURLOPT_POST,1);
            curl_setopt($rest,CURLOPT_POSTFIELDS,$push_payload);
            curl_setopt($rest, CURLOPT_RETURNTRANSFER, TRUE);
            curl_setopt($rest,CURLOPT_HTTPHEADER,
                    array("X-Parse-Application-Id: " . $APPLICATION_ID,
                            "X-Parse-REST-API-Key: " . $REST_API_KEY,
                            "Content-Type: application/json"));

            // Execute post
            $response = curl_exec($rest);
            //pr($rest);
            //pr($response);
            // Close connection
            curl_close($rest);
            //exit;
            if (!$response) {
                throw new Exception("CURL execution fails..!");
            }

            $success = 1;
            $message = "Push notification send successfully..!";
        } catch (Exception $e) {
            $success = 0;
            $message = $e->getMessage();
        }

        return $success;
    }

Let me know in case further details require.

**Comment by grantland** [comment]
It's not entirely clear whether this is a SDK issue, so we'll need some more information:

* The Parse SDK version you're using.
* Whether you're using GCM or PPNS. You can find out by adding `Parse.setLogLevel(Parse.LOG_LEVEL_VERBOSE)` and looking for "Using <GCM or PPNS> for push" in logcat.
* Whether or not the device is getting duplicate notifications. You can find out by using your own `BroadcastReceiver` in place of `com.parse.GcmBroadcastReceiver` and overriding `onReceive` with a log message that outputs the data in the extras `Bundle`.
* Whether or not this repros when you use the REST API directly (without your serverside PHP script).

**Comment by hbabhishek** [comment]
Following is the information requested by you:

- We are using Parse SDK version 1.7.1
- We are using GCM
- The duplicate notification issue is generated randomly so we can not reproduce it at our end but one user has sent us the screen shot attached above complaining about the duplicate notifications.

**Comment by grantland** [comment]
Thanks, but we'll need all requested information in order to verify this is actually a SDK issue or not.

**Comment by hbabhishek** [comment]
Hi I have already given all information requested by you. Can you please let me know what other information you need?

**Comment by grantland** [comment]
* Whether or not the device is getting duplicate notifications from GCM. You can find out by using your own `BroadcastReceiver` in place of `com.parse.GcmBroadcastReceiver` and overriding `onReceive` with a log message that outputs the data in the extras `Bundle`.
* Whether or not this repros when you use the REST API directly (without your serverside PHP script).

**Comment by hbabhishek** [comment]
I tried the suggestions you gave above. Everything works perfect in test environment. But in production environment some users are facing this issue.

**Comment by grantland** [comment]
Unfortunately without this information or reproduction steps we aren't able to provide a solution as we won't be able to determine if the problem is due to the SDK, backend, or application.

**Comment by odbol** [comment]
I'm having this same problem. Followed the Getting Started guide exactly. Updating to 1.10.3 SDK did not help. 

If you check our "Tappur" app's data table, you'll see many duplicate _Installation objects. They are all the exact same, except for the "timestamp" column. This seems to be pushing multiple notifications to the same device. 
<img width="1170" alt="screen shot 2015-10-20 at 3 22 10 pm" src="https://cloud.githubusercontent.com/assets/550985/10622800/9765f494-773e-11e5-8737-43a4eee74493.png">
 
Many people are having this same issue:

https://www.parse.com/questions/duplicate-installations-are-counted-for-push-notifications
https://parse.com/questions/parse-sends-multiple-notifications-using-gcm
http://stackoverflow.com/questions/24393027/getting-multiple-notifications-per-push

It's really simple to reproduce:

1. Follow Getting Started guide for Android.
2. Install the app
3. Uninstall the app.
4. Install the app again.
5. Repeat 2-4 as many times as you want.
6. Send push notification
7. You'll get multiple notifications on the same device.

**Comment by bnham** [comment]
This occurs because when you uninstall an app and then reinstall an app, GCM registers for another reg id which is also associated with your app. Eventually these get cleaned up after you send a push because the GCM server will tell us *after* sending the push that the two different reg ids are actually associated with the same reg id. We then update our server to delete the old outdated reg id, so the next push won't result in a duplicated push. This is all normal behavior for an app that uses GCM.

In any case, this shouldn't be an issue anymore in SDK 1.9.4+ because we added a deduping step to the push receiver:

https://github.com/ParsePlatform/Parse-SDK-Android/blob/master/Parse/src/main/java/com/parse/PushRouter.java#L142-L144

Please upgrade your SDK to a later version.

**Comment by bnham** [comment]
What you pasted is not a snapshot of the data in your app. That's a list of API events. What you pasted is basically showing the same device made multiple requests to our API server. If you go to the actual data browser (e.g. https://www.parse.com/apps/$YOUR_APP_NAME/collections#class/_Installation) you would see that the installation objects aren't duplicated.

Secondly, all those questions pertain to Parse SDK <1.9.4. In Parse SDK 1.9.4 we added client-side deduping of GCM pushes. Long story short, duplicate pushes can occur via GCM because of the way GCM lazily invalidates old registration ids associated with uninstalled apps. This is not a permanent state as the GCM server tells us after we send the push that the reg ids are actually referring to the same device, so we then go and delete the old installation object associated with the old reg id server-side. But because we got enough complaints about this behavior we now do client-side deduping of GCM pushes:

https://github.com/ParsePlatform/Parse-SDK-Android/blob/master/Parse/src/main/java/com/parse/PushRouter.java#L28-L30
https://github.com/ParsePlatform/Parse-SDK-Android/blob/master/Parse/src/main/java/com/parse/PushRouter.java#L142-L144

If you still see this behavior in 1.10.3, then attach the contents of the "push" file from your app data directory (this is a path like `/data/data/com.yourapp.id/app_Parse/push`). And also let us know the push id of the push you think was duplicated (the url when you click on details of your push on the website, e.g. https://www.parse.com/apps/$YOUR_APP_NAME/push_notifications/abCde0123ba).
